design/design-doc-agent.md

```diff
@@ -0,0 +1,665 @@
+  proxy of workload CRs. You're losing on some value that these models provide
+  because of that proxy and in some cases it could be functionally detrimental.
+
+Surely, it has its own advantages as well. For example, you can manage all of
+your apps from single point via `KubernetesApplication`s targeting the right
+clusters. But as we see more usage patterns, we're convinced that the current
+mechanics do not provide the experience users would like to have.
+
+## Proposal
+
+In order to preserve the central infrastructure management ability while
+alleviating the issues above, we will change our approach from push-based one to
+a pull-based one where applications, and their requirements are deployed into
+the remote cluster, and they request the infrastructure from a central cluster
+and pull the necessary credentials.
+
+Since having this logic in the applications themselves wouldn't be a good UX, we
+will have an agent that you will need to deploy into your remote cluster for
+doing the heavy-lifting for you. There are several technical problems to be
+solved in order to make the experience smooth. Overall, the goal is that we want
+to keep the UX of local mode for application operators while keeping the power
+of centralized infrastructure management for platform operators. For reference,
+here is an example local mode experience we'd like to have for the remote mode
+as well:
+
+```yaml
+apiVersion: common.crossplane.io/v1alpha1
+kind: MySQLInstanceRequirement
+metadata:
+  name: sqldb
+  namespace: default
+spec:
+  version: "5.7"
+  storageGB: 20
+  writeConnectionSecretToRef:
+    name: sql-creds
+---
+apiVersion: v1
+kind: Pod
+metadata:
+  name: wp
+  namespace: default
+spec:
+containers:
+- name: wordpress
+    image: "wordpress:4.6.1-apache"
+    env:
+    - name: WORDPRESS_DB_PASSWORD
+    valueFrom:
+        secretKeyRef:
+          name: sql-creds
+          key: password
+```
+
+The agent will be a Kubernetes controller running in the remote cluster and
+watching all `*Requirement` types. Next sections will talk about the
+implementation and user experience we'd like to have.
+
+### Synchronization
+
+In local mode, users directly interact with what's published by the platform
+team, which is `*Requirement` types and consume the infrastructure by mounting
+the secret whose name they specified on the `*Requirement` custom resource. To
+keep this experience, we need to have a synchronization loop for the following
+resources:
+
+* Pull
+  * `CustomResourceDefinition`s of all types that we want the applications to be
+    able to manage and view:
+    * Requirements that are published via `InfrastructurePublication`s. The
+      source of truth will be the remote cluster.
+    * `InfrastructureDefinition`, `InfrastructurePublication` and `Composition`.
+  * `Composition`s to discover what's available to choose. CRs of this type will
+    be read-only, and the source of truth will be the central cluster.
```

Comment:
How should we enforce the read-only nature of Compositions, InfraDefs, and InfraPubs?

Reply:
The agent will blindly `Update` those resources in the remote cluster, meaning even if the user changes a value, it will be overridden right away.

Reply:
Got it. I guess RBAC could be used as well to ensure that no one has access to update them in the first place.